mlflow/genai/evaluation/context.py

```diff
@@ -0,0 +1,156 @@
+    """
+    A context that does nothing.
+    """
+
+    def get_mlflow_experiment_id(self) -> Optional[str]:
+        raise NotImplementedError("Context is not set")
+
+    def get_mlflow_run_id(self) -> Optional[str]:
+        raise NotImplementedError("Context is not set")
+
+    def get_user_name(self) -> Optional[str]:
+        raise NotImplementedError("Context is not set")
+
+
+class RealContext(Context):
+    """
+    Context for eval execution.
+
+    NOTE: This class is not covered by unit tests and is meant to be tested through
+    smoke tests that run this code on an actual Databricks cluster.
+    """
+
+    @classmethod
+    def _get_dbutils(cls):
+        """
+        Returns an instance of dbutils.
+        """
+        try:
+            from databricks.sdk.runtime import dbutils
+
+            return dbutils
+        except ImportError:
+            import IPython
+
+            dbutils = IPython.get_ipython().user_ns["dbutils"]
+        return dbutils
+
+    def __init__(self):
+        self._dbutils = self._get_dbutils()
+        self._run_id = None
+        try:
+            self._notebook_context = self._dbutils.entry_point.getDbutils().notebook().getContext()
+        except Exception:
+            self._notebook_context = None
+
+    def get_mlflow_experiment_id(self) -> Optional[str]:
+        # Note `_get_experiment_id` is thread-safe
+        return mlflow.tracking.fluent._get_experiment_id()
+
+    def get_mlflow_run_id(self) -> Optional[str]:
+        """
+        Gets the MLflow run_id the evaluation harness is running under.
+
+        Warning: This run_id may not be active. This happens when `get_mlflow_run_id` is called from
+        a different thread than the one that started the MLflow run.
+        """
+        # First check if a run ID is specified explicitly by the parent thread
+        if self._run_id:
+            return self._run_id
+
+        # Otherwise fall back to the active run in the current thread
+        if run := mlflow.active_run():
+            return run.info.run_id
+
+        return None
+
+    def set_mlflow_run_id(self, run_id: str) -> None:
+        """
+        Set the MLflow run ID explicitly.
+
+        This method should be called when running code in a different thread than the one that
+        started the MLflow run. It sets the run ID in a thread-local variable so that it can be
+        accessed from the thread.
+        """
+        self._run_id = run_id
+
+    def get_user_name(self) -> Optional[str]:
+        try:
+            return self._notebook_context.userName().get()
+        except Exception:
+            return None
+
+
+# Context is a singleton.
+_context_singleton = NoneContext()
+
+
+def context_is_active() -> bool:
+    """
+    Check if a context is active.
+    """
+    return not isinstance(get_context(), NoneContext)
+
+
+def get_context() -> Context:
+    """
+    Get the context.
+    """
+    return _context_singleton or NoneContext()
```

Comment:
```suggestion
    return _context_singleton
```

`_context_singleton` is always truthy.